File: bindings/ruby/stores/postgres/lib/prolly/store/postgres.rb
# frozen_string_literal: true

require 'pg'
require 'prolly'

module Prolly
  class PostgresRemoteStore < ForeignRemoteStore
    CREATE_SCHEMA = <<~SQL
      CREATE TABLE IF NOT EXISTS prolly_nodes (cid bytea PRIMARY KEY, node bytea NOT NULL);
      CREATE TABLE IF NOT EXISTS prolly_hints (namespace bytea NOT NULL, key bytea NOT NULL, value bytea NOT NULL, PRIMARY KEY (namespace, key));
      CREATE TABLE IF NOT EXISTS prolly_roots (name bytea PRIMARY KEY, manifest bytea NOT NULL);
    SQL
    UPSERT_NODE = 'INSERT INTO prolly_nodes VALUES ($1::bytea, $2::bytea) ON CONFLICT(cid) DO UPDATE SET node=excluded.node'
    UPSERT_HINT = 'INSERT INTO prolly_hints VALUES ($1::bytea, $2::bytea, $3::bytea) ON CONFLICT(namespace,key) DO UPDATE SET value=excluded.value'
    UPSERT_ROOT = 'INSERT INTO prolly_roots VALUES ($1::bytea, $2::bytea) ON CONFLICT(name) DO UPDATE SET manifest=excluded.manifest'
    LOCK_ROOT = "SELECT pg_advisory_xact_lock(hashtextextended(encode($1::bytea, 'hex'), 0))"

    def initialize(connection)
      raise ArgumentError, 'connection must be PG::Connection' unless connection.is_a?(PG::Connection)

      @connection = connection
      @mutex = Mutex.new
      @closed = false
    end

    def initialize_schema = synchronize { @connection.exec(CREATE_SCHEMA) }
    def close = @closed = true

    def descriptor
      capabilities = StoreCapabilitiesRecord.new(
        native_batch_reads: false, atomic_batch_writes: true, node_scan: true,
        hints: true, atomic_nodes_and_hint: true, root_scan: true,
        root_compare_and_swap: true, transactions: true, read_parallelism: 16
      )
      limits = StoreLimitsRecord.new(
        max_batch_read_items: nil, max_batch_write_items: nil,
        max_transaction_operations: nil, max_node_bytes: nil
      )
      value = StoreDescriptorRecord.new(
        protocol_major: 1, adapter_name: 'postgres-v1', provider: 'postgresql',
        schema_version: 1, capabilities: capabilities, limits: limits
      )
      StoreDescriptorResultRecord.new(value: value, error: nil)
    end

    def get_node(cid) = optional_result(query('SELECT node FROM prolly_nodes WHERE cid = $1::bytea', cid))
    def put_node(cid, value) = write(UPSERT_NODE, cid, value)
    def delete_node(cid) = write('DELETE FROM prolly_nodes WHERE cid = $1::bytea', cid)

    def batch_nodes(operations)
      transaction do
        operations.each do |item|
          item.value.present ? execute(UPSERT_NODE, item.key, item.value.value) : execute('DELETE FROM prolly_nodes WHERE cid = $1::bytea', item.key)
        end
      end
      unit
    rescue StandardError => error
      UnitResultRecord.new(error: store_error(error))
    end

    def batch_get_nodes_ordered(cids)
      values = synchronize { cids.map { |cid| optional(query_unlocked('SELECT node FROM prolly_nodes WHERE cid = $1::bytea', cid)) } }
      OptionalBytesListResultRecord.new(values: values, error: nil)
    rescue StandardError => error
      OptionalBytesListResultRecord.new(values: [], error: store_error(error))
    end

    def list_node_cids
      values = synchronize { @connection.exec('SELECT cid FROM prolly_nodes ORDER BY cid').map { |row| decode(row['cid']) } }
      BytesListResultRecord.new(values: values, error: nil)
    rescue StandardError => error
      BytesListResultRecord.new(values: [], error: store_error(error))
    end

    def get_hint(namespace, key) = optional_result(query('SELECT value FROM prolly_hints WHERE namespace = $1::bytea AND key = $2::bytea', namespace, key))
    def put_hint(namespace, key, value) = write(UPSERT_HINT, namespace, key, value)

    def batch_put_nodes_with_hint(nodes, namespace, key, value)
      transaction do
        nodes.each { |node| execute(UPSERT_NODE, node.key, node.value) }
        execute(UPSERT_HINT, namespace, key, value)
      end
      unit
    rescue StandardError => error
      UnitResultRecord.new(error: store_error(error))
    end

    def get_root_manifest(name) = optional_result(query('SELECT manifest FROM prolly_roots WHERE name = $1::bytea', name))
    def put_root_manifest(name, manifest) = write(UPSERT_ROOT, name, manifest)
    def delete_root_manifest(name) = write('DELETE FROM prolly_roots WHERE name = $1::bytea', name)

    def compare_and_swap_root_manifest(name, expected, replacement)
      synchronize do
        applied = false
        current = nil
        transaction_unlocked do
          execute(LOCK_ROOT, name)
          current = query_unlocked('SELECT manifest FROM prolly_roots WHERE name = $1::bytea FOR UPDATE', name)
          if matches?(current, expected)
            write_root(name, replacement)
            current = replacement.present ? replacement.value : nil
            applied = true
          end
        end
        RootCasResultRecord.new(applied: applied, current: optional(current), error: nil)
      end
    rescue StandardError => error
      RootCasResultRecord.new(applied: false, current: optional(nil), error: store_error(error))
    end

    def list_root_manifests
      values = synchronize do
        @connection.exec('SELECT name, manifest FROM prolly_roots ORDER BY name').map do |row|
          NamedBytesRecord.new(name: decode(row['name']), value: decode(row['manifest']))
        end
      end
      NamedBytesListResultRecord.new(values: values, error: nil)
    rescue StandardError => error
      NamedBytesListResultRecord.new(values: [], error: store_error(error))
    end

    def commit_transaction(nodes, conditions, roots)
      conflict = nil
      synchronize do
        transaction_unlocked do
          conditions.map(&:name).uniq.sort.each { |name| execute(LOCK_ROOT, name) }
          conditions.each do |condition|
            current = query_unlocked('SELECT manifest FROM prolly_roots WHERE name = $1::bytea FOR UPDATE', condition.name)
            next if matches?(current, condition.expected)

            conflict = StoreTransactionConflictRecord.new(
              name: condition.name, expected: condition.expected, current: optional(current)
            )
            break
          end
          unless conflict
            nodes.each do |item|
              item.value.present ? execute(UPSERT_NODE, item.key, item.value.value) : execute('DELETE FROM prolly_nodes WHERE cid = $1::bytea', item.key)
            end
            roots.each { |root| write_root(root.name, root.replacement) }
          end
        end
      end
      TransactionResultRecord.new(applied: conflict.nil?, conflict: conflict, error: nil)
    rescue StandardError => error
      TransactionResultRecord.new(applied: false, conflict: nil, error: store_error(error))
    end

    private

    def synchronize(&block)
      raise 'PostgreSQL store is closed' if @closed
      @mutex.synchronize(&block)
    end

    def encode(value) = @connection.escape_bytea(value.b)
    def decode(value) = PG::Connection.unescape_bytea(value).b
    def execute(sql, *values) = @connection.exec_params(sql, values.map { |value| encode(value) })
    def query(sql, *values) = synchronize { query_unlocked(sql, *values) }

    def query_unlocked(sql, *values)
      result = execute(sql, *values)
      result.ntuples.zero? ? nil : decode(result.getvalue(0, 0))
    end

    def write(sql, *values)
      synchronize { execute(sql, *values) }
      unit
    rescue StandardError => error
      UnitResultRecord.new(error: store_error(error))
    end

    def transaction(&block) = synchronize { transaction_unlocked(&block) }

    def transaction_unlocked
      @connection.exec('BEGIN')
      begin
        value = yield
        @connection.exec('COMMIT')
        value
      rescue Exception
        @connection.exec('ROLLBACK') rescue nil
        raise
      end
    end

    def optional(value) = OptionalBytesRecord.new(present: !value.nil?, value: value || ''.b)
    def optional_result(value) = OptionalBytesResultRecord.new(value: optional(value), error: nil)
    def unit = UnitResultRecord.new(error: nil)
    def matches?(current, expected) = expected.present ? current == expected.value : current.nil?

    def write_root(name, replacement)
      replacement.present ? execute(UPSERT_ROOT, name, replacement.value) : execute('DELETE FROM prolly_roots WHERE name = $1::bytea', name)
    end

    def store_error(error)
      StoreErrorRecord.new(
        code: 'internal', message: 'PostgreSQL provider operation failed', retryable: false,
        provider_code: error.respond_to?(:result) ? error.result&.error_field(PG::Result::PG_DIAG_SQLSTATE) : nil
      )
    end
  end
end

File: bindings/go/stores/postgres/postgres.go
package postgres

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"strings"
	"sync/atomic"

	prolly "build.crab/prolly-go"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgconn"
	"github.com/jackc/pgx/v5/pgxpool"
)

type Options struct {
	AdapterName     string
	ReadParallelism uint32
}

type Store struct {
	pool    *pgxpool.Pool
	options Options
	owned   bool
	closed  atomic.Bool
}

func New(pool *pgxpool.Pool, options Options) *Store {
	return &Store{pool: pool, options: normalizeOptions(options)}
}

func Open(ctx context.Context, connectionString string, options Options) (*Store, error) {
	pool, err := pgxpool.New(ctx, connectionString)
	if err != nil {
		return nil, pgError("open", err)
	}
	store := New(pool, options)
	store.owned = true
	return store, nil
}

func normalizeOptions(options Options) Options {
	if strings.TrimSpace(options.AdapterName) == "" {
		options.AdapterName = "postgres-v1"
	}
	if options.ReadParallelism == 0 {
		options.ReadParallelism = 16
	}
	return options
}

func (s *Store) InitializeSchema(ctx context.Context) error {
	if err := s.ready(ctx); err != nil {
		return err
	}
	for _, statement := range strings.Split(Schema, ";") {
		if strings.TrimSpace(statement) == "" {
			continue
		}
		if _, err := s.pool.Exec(ctx, statement); err != nil {
			return pgError("initialize_schema", err)
		}
	}
	return nil
}

func (s *Store) Close() error {
	if s == nil || s.closed.Swap(true) || !s.owned || s.pool == nil {
		return nil
	}
	s.pool.Close()
	return nil
}

func (s *Store) Descriptor(ctx context.Context) (prolly.StoreDescriptor, error) {
	if err := s.ready(ctx); err != nil {
		return prolly.StoreDescriptor{}, err
	}
	return prolly.StoreDescriptor{
		ProtocolMajor: prolly.StoreProtocolMajor, AdapterName: s.options.AdapterName, Provider: "postgresql", SchemaVersion: 1,
		Capabilities: prolly.StoreCapabilities{
			NativeBatchReads: true, AtomicBatchWrites: true, NodeScan: true, Hints: true,
			AtomicNodesAndHint: true, RootScan: true, RootCompareAndSwap: true,
			Transactions: true, ReadParallelism: s.options.ReadParallelism,
		},
	}, nil
}

func (s *Store) GetNode(ctx context.Context, key []byte) (prolly.OptionalBytes, error) {
	if err := s.ready(ctx); err != nil {
		return prolly.OptionalBytes{}, err
	}
	return queryOptional(ctx, s.pool, selectNode, key)
}

func (s *Store) PutNode(ctx context.Context, key, value []byte) error {
	if err := s.ready(ctx); err != nil {
		return err
	}
	_, err := s.pool.Exec(ctx, upsertNode, key, value)
	return pgError("put_node", err)
}

func (s *Store) DeleteNode(ctx context.Context, key []byte) error {
	if err := s.ready(ctx); err != nil {
		return err
	}
	_, err := s.pool.Exec(ctx, deleteNode, key)
	return pgError("delete_node", err)
}

func (s *Store) BatchNodes(ctx context.Context, mutations []prolly.NodeMutation) error {
	if err := s.ready(ctx); err != nil {
		return err
	}
	return s.withTx(ctx, "batch_nodes", func(tx pgx.Tx) error { return applyNodes(ctx, tx, mutations) })
}

func (s *Store) PublishNodes(ctx context.Context, publication prolly.NodePublication) error {
	return prolly.PublishNodesWithGeneralPath(ctx, s, publication)
}

func (s *Store) BatchGetNodesOrdered(ctx context.Context, keys [][]byte) ([]prolly.OptionalBytes, error) {
	if err := s.ready(ctx); err != nil {
		return nil, err
	}
	result := make([]prolly.OptionalBytes, len(keys))
	for index, key := range keys {
		value, err := queryOptional(ctx, s.pool, selectNode, key)
		if err != nil {
			return nil, err
		}
		result[index] = value
	}
	return result, nil
}

func (s *Store) ListNodeCIDs(ctx context.Context) ([][]byte, error) {
	if err := s.ready(ctx); err != nil {
		return nil, err
	}
	rows, err := s.pool.Query(ctx, `SELECT cid FROM prolly_nodes ORDER BY cid`)
	if err != nil {
		return nil, pgError("list_nodes", err)
	}
	defer rows.Close()
	var result [][]byte
	for rows.Next() {
		var value []byte
		if err := rows.Scan(&value); err != nil {
			return nil, pgError("list_nodes_scan", err)
		}
		result = append(result, clone(value))
	}
	return result, pgError("list_nodes_rows", rows.Err())
}

func (s *Store) GetHint(ctx context.Context, namespace, key []byte) (prolly.OptionalBytes, error) {
	if err := s.ready(ctx); err != nil {
		return prolly.OptionalBytes{}, err
	}
	return queryOptional(ctx, s.pool, selectHint, namespace, key)
}

func (s *Store) PutHint(ctx context.Context, namespace, key, value []byte) error {
	if err := s.ready(ctx); err != nil {
		return err
	}
	_, err := s.pool.Exec(ctx, upsertHint, namespace, key, value)
	return pgError("put_hint", err)
}

func (s *Store) BatchPutNodesWithHint(ctx context.Context, nodes []prolly.NodeEntry, namespace, key, value []byte) error {
	if err := s.ready(ctx); err != nil {
		return err
	}
	return s.withTx(ctx, "batch_nodes_hint", func(tx pgx.Tx) error {
		for _, node := range nodes {
			if _, err := tx.Exec(ctx, upsertNode, node.Key, node.Value); err != nil {
				return pgError("batch_node", err)
			}
		}
		_, err := tx.Exec(ctx, upsertHint, namespace, key, value)
		return pgError("batch_hint", err)
	})
}

func (s *Store) GetRootManifest(ctx context.Context, name []byte) (prolly.OptionalBytes, error) {
	if err := s.ready(ctx); err != nil {
		return prolly.OptionalBytes{}, err
	}
	return queryOptional(ctx, s.pool, selectRoot, name)
}

func (s *Store) PutRootManifest(ctx context.Context, name, manifest []byte) error {
	if err := s.ready(ctx); err != nil {
		return err
	}
	_, err := s.pool.Exec(ctx, upsertRoot, name, manifest)
	return pgError("put_root", err)
}

func (s *Store) DeleteRootManifest(ctx context.Context, name []byte) error {
	if err := s.ready(ctx); err != nil {
		return err
	}
	_, err := s.pool.Exec(ctx, deleteRoot, name)
	return pgError("delete_root", err)
}

func (s *Store) CompareAndSwapRootManifest(ctx context.Context, name []byte, expected, replacement prolly.OptionalBytes) (prolly.RootCASResult, error) {
	if err := s.ready(ctx); err != nil {
		return prolly.RootCASResult{}, err
	}
	var result prolly.RootCASResult
	err := s.withTx(ctx, "root_cas", func(tx pgx.Tx) error {
		if _, err := tx.Exec(ctx, `LOCK TABLE prolly_roots IN SHARE ROW EXCLUSIVE MODE`); err != nil {
			return pgError("lock_roots", err)
		}
		current, err := queryOptional(ctx, tx, selectRoot, name)
		if err != nil {
			return err
		}
		if !optionalEqual(current, expected) {
			result.Current = current
			return nil
		}
		if err := writeRoot(ctx, tx, name, replacement); err != nil {
			return err
		}
		result = prolly.RootCASResult{Applied: true, Current: replacement.Clone()}
		return nil
	})
	return result, err
}

func (s *Store) ListRootManifests(ctx context.Context) ([]prolly.NamedStoreRoot, error) {
	if err := s.ready(ctx); err != nil {
		return nil, err
	}
	rows, err := s.pool.Query(ctx, `SELECT name, manifest FROM prolly_roots ORDER BY name`)
	if err != nil {
		return nil, pgError("list_roots", err)
	}
	defer rows.Close()
	var result []prolly.NamedStoreRoot
	for rows.Next() {
		var name, manifest []byte
		if err := rows.Scan(&name, &manifest); err != nil {
			return nil, pgError("list_roots_scan", err)
		}
		result = append(result, prolly.NamedStoreRoot{Name: clone(name), Manifest: clone(manifest)})
	}
	return result, pgError("list_roots_rows", rows.Err())
}

func (s *Store) CommitTransaction(ctx context.Context, nodes []prolly.NodeMutation, conditions []prolly.RootCondition, roots []prolly.RootWrite) (prolly.StoreTransactionResult, error) {
	if err := s.ready(ctx); err != nil {
		return prolly.StoreTransactionResult{}, err
	}
	var result prolly.StoreTransactionResult
	err := s.withTx(ctx, "transaction", func(tx pgx.Tx) error {
		if _, err := tx.Exec(ctx, `LOCK TABLE prolly_roots IN SHARE ROW EXCLUSIVE MODE`); err != nil {
			return pgError("lock_roots", err)
		}
		for _, condition := range conditions {
			current, err := queryOptional(ctx, tx, selectRoot, condition.Name)
			if err != nil {
				return err
			}
			if !optionalEqual(current, condition.Expected) {
				result.Conflict = &prolly.StoreTransactionConflict{Name: clone(condition.Name), Expected: condition.Expected.Clone(), Current: current}
				return nil
			}
		}
		if err := applyNodes(ctx, tx, nodes); err != nil {
			return err
		}
		for _, root := range roots {
			if err := writeRoot(ctx, tx, root.Name, root.Replacement); err != nil {
				return err
			}
		}
		result.Applied = true
		return nil
	})
	return result, err
}

type queryer interface {
	QueryRow(context.Context, string, ...any) pgx.Row
}
type execer interface {
	Exec(context.Context, string, ...any) (pgconn.CommandTag, error)
}

func queryOptional(ctx context.Context, target queryer, query string, args ...any) (prolly.OptionalBytes, error) {
	var value []byte
	err := target.QueryRow(ctx, query, args...).Scan(&value)
	if errors.Is(err, pgx.ErrNoRows) {
		return prolly.MissingBytes(), nil
	}
	if err != nil {
		return prolly.OptionalBytes{}, pgError("query", err)
	}
	return prolly.PresentBytes(value), nil
}

func applyNodes(ctx context.Context, tx pgx.Tx, mutations []prolly.NodeMutation) error {
	for _, mutation := range mutations {
		var err error
		if mutation.Value.Present {
			_, err = tx.Exec(ctx, upsertNode, mutation.Key, mutation.Value.Value)
		} else {
			_, err = tx.Exec(ctx, deleteNode, mutation.Key)
		}
		if err != nil {
			return pgError("node_mutation", err)
		}
	}
	return nil
}

func writeRoot(ctx context.Context, target execer, name []byte, value prolly.OptionalBytes) error {
	var err error
	if value.Present {
		_, err = target.Exec(ctx, upsertRoot, name, value.Value)
	} else {
		_, err = target.Exec(ctx, deleteRoot, name)
	}
	return pgError("write_root", err)
}

func (s *Store) withTx(ctx context.Context, operation string, call func(pgx.Tx) error) error {
	tx, err := s.pool.BeginTx(ctx, pgx.TxOptions{})
	if err != nil {
		return pgError(operation+"_begin", err)
	}
	defer tx.Rollback(ctx)
	if err := call(tx); err != nil {
		return err
	}
	return pgError(operation+"_commit", tx.Commit(ctx))
}

func optionalEqual(left, right prolly.OptionalBytes) bool {
	return left.Present == right.Present && (!left.Present || bytes.Equal(left.Value, right.Value))
}
func clone(value []byte) []byte { return append([]byte(nil), value...) }

func (s *Store) ready(ctx context.Context) error {
	if ctx != nil && ctx.Err() != nil {
		return ctx.Err()
	}
	if s == nil || s.pool == nil {
		return &prolly.StoreError{Code: "invalid_store", Message: "PostgreSQL pool is nil"}
	}
	if s.closed.Load() {
		return &prolly.StoreError{Code: "closed", Message: "PostgreSQL store is closed"}
	}
	return nil
}

func pgError(operation string, err error) error {
	if err == nil {
		return nil
	}
	retryable := false
	providerCode := ""
	var pgErr *pgconn.PgError
	if errors.As(err, &pgErr) {
		providerCode = pgErr.Code
		retryable = strings.HasPrefix(pgErr.Code, "08") || pgErr.Code == "40001" || pgErr.Code == "40P01" || pgErr.Code == "55P03"
	}
	return &prolly.StoreError{Code: "postgresql", Message: fmt.Sprintf("%s: %v", operation, err), Retryable: retryable, ProviderCode: providerCode, Cause: err}
}

var _ prolly.RemoteStore = (*Store)(nil)
